bundler/lib/bundler/definition.rb

```diff
@@ -417,6 +420,7 @@ def ensure_equivalent_gemfile_and_lockfile(explicit_flag = false)
     def validate_runtime!
       validate_ruby!
       validate_platforms!
+      validate_plugins!
```

Comment:
I'm not sure what you're hoping for here. The validate_plugins! piece is checking that all plugins referenced in the Gemfile actually exist. It needs to run for any bundle command (even just bundle/setup, from the app) because if the plugins index references a gem (by path!) that has been removed for any reason (referencing a system gem that has been uninstalled; referencing a path gem that the directory was removed, etc.), you get an indecipherable loading error when Bundler tries to load the plugin. This includes even if you're running `bundle check`.

Reply:
I'm just curious, since we don't do the same validation for regular specs here, how Bundler behaves in this situation for regular gems, since we have not heard about such errors regarding gems. I think we automatically reinstall them if missing in some situations, and give a useful error when materializing otherwise (no upfront validation).

I mentioned `bundle check` because it was the only thing that failed in specs when commenting this out.

Reply:
I could probably add more specs around this. `bundle install`, or even simply `require "bundler/setup"` in the app will fail with a near-indecipherable exception about not being able to load plugins.rb without this.

Reply:
I added an additional check in the spec for this that `bundle exec rake -T` also fails gracefully, and ensured that without this line, it indeed fails with a cryptic error about a path not existing:
```
(`/Users/cody/src/rubygems/bundler/tmp/1/gems/system/gems/bundler-2.6.0.dev/lib/bundler/source/path.rb:197:in `load_spec_files': The path `/Users/cody/src/rubygems/bundler/tmp/1/libs/ga-plugin-1.0` does not exist. (Bundler::PathError)
```

Reply:
Okay, let me give you a more real world example. Say I'm working on a project, called `canvas-lms`. My Gemfile currently looks like this at the top:

```ruby
source "https://rubygems.org/"

plugin "bundler-multilock", "1.3.4"
# if the plugin is not installed yet, we can't continue, since the Gemfile uses DSL methods added by the plugin
return unless Plugin.installed?("bundler-multilock")
```

_without_ `validate_plugins!`, if I do a fresh checkout of the repository, and am not really paying attention and just run something like `bin/rails db:migrate` to start setting things up (or more simply run `bundle check`), it will _not_ run the plugin install phase (which is only part of `bundle install` and friends). Assuming I have the `bundler-multilock` gem installed on my system (from use in another project), it will happily output `The Gemfile's dependencies are satisfied`. BUT, the plugin never got registered in this bundle, so it never ran, and the Gemfile exited early, and my Gemfile.lock gets completely replaced with this:

```
GEM
  remote: https://rubygems.org/
  specs:
    bundler-multilock (1.3.2)
      bundler (>= 2.4.19, < 2.6)

PLATFORMS
  aarch64-linux
  arm64-darwin
  ruby
  x86_64-darwin
  x86_64-linux

DEPENDENCIES
  bundler-multilock (~> 1.3)

BUNDLED WITH
   2.6.0.dev
```

A poor outcome indeed. With the `validate_plugins!` check, it outputs `Plugin bundler-multilock (~> 1.3) is not installed` before it tries to evaluate the Gemfile in its primary pass. One could argue the Gemfile should raise an error instead of simply returning in its own check for the plugin being installed (like `raise "Please run `bundle install`" unless Plugin.installed?("bundler-multilock")`), resulting in an error like this:

```
$ bundle check

[!] There was an error parsing `Gemfile`: Please run `bundle install`. Bundler cannot continue.

 #  from /Users/cody/src/canvas-lms/Gemfile:18
 #  -------------------------------------------
 #  plugin "bundler-multilock", "~> 1.3"
 >  raise "Please run `bundle install`" unless Plugin.installed?("bundler-multilock")
 #  
 #  -------------------------------------------
```

The problem is that this you can _never_ resolve this, since the error is also thrown during the plugin-only phase of evaluating the Gemfile.

A slightly different scenario, but with the exact same outcome (Gemfile.lock getting clobbered) is if you share a single working copy among multiple logical OSes (like a mounted volume inside docker where you do most development, but run rubocop on the host OS as a commit hook).  In this case, the plugin index will reference a path that's not valid inside of Docker, so will treat the plugin as if it's not installed.

I think the core reason `validate_plugins!` is necessary is a combination of having the two-phase install process (as discussed in the other still open thread), and that not all commands run that first plugin-only phase (where the normal "are all gems available, and by gems, I mean plugins") where a broken plugin index would be caught. Perhaps that's evidence that a regular `bundle/setup` should run the first phase? But is that something we want to address with this PR? I'm hesitant to add even more to this PR, especially since I worry about performance implications of if that would mean loading install infrastructure when bundler normally doesn't.

Reply:
Just to understand the context, since this is new to me, is this `return unless Plugin.installed?("bundler-multilock")` line something necessary for many plugins, or is it something specific to your `bundler-multilock` plugin, or related to the changes in this PR? I thought in order to use a plugin, the `plugin` DSL was enough. Is this something we can "fix" inside Bundler instead, so it's not necessary?

The experience you mention about the lockfile getting replaced with just the plugin and the plugin not running is indeed not great, but it's not super surprising after seeing the `return unless Plugin.installed?("bundler-multilock")` line in the `Gemfile`. I guess it'd be something made worse after this PR if we don't add this validation, since we did not lock plugins before, so good reason to do something about it. But I wonder if we can do something to avoid having to include that kind of code in the `Gemfile`.

Also, my comment before still stands. I think this degrades the experience at least in the particular case of the spec that you added, namely, using `plugin 'ga-plugin', :path => "#{lib_path("ga-plugin-1.0")}"` and then renaming/removing `lib_path("ga-plugin-1.0")`, since the current error at least mentions that the path does not exist while the new error mentions that the plugin "is not installed", which does not make a lot of sense because as far as I understand "path plugins" are not really installed, like regular "path gems". And for this particular case, any improvement in the error message would benefit both `gem` and `plugin` usages.

Reply:
Also, I wonder if https://github.com/rubygems/rubygems/pull/7639 will improve the situation?

Reply:
#7639 is definitely related! That was the original situation I encountered that I added this for. I'm a little surprised this PR didn't resolve her issue.



> is it something specific to your `bundler-multilock` plugin

Yes and no. Using `return` in the Gemfile is my solution to this class of problem. But this class of problem would apply to any plugin that extends the Gemfile DSL. bootboot avoids the problem by not actually modifying syntax of the DSL, but just tweaking ENV vars which are still valid Ruby syntax regardless of if the plugin has loaded yet. bundler-override addresses it by having the user [add a line](https://github.com/tarnowsc/bundler-override?tab=readme-ov-file#usage) to the Gemfile that immediately loads the plugin, even in the plugin-only phase of parsing - _and_ adding a `rescue nil` (😱) in that line to ignore the error that occurs when the plugin isn't installed yet (though this seems to me that the initial install would always be broken, because then it will fail later on the `override` call).

> Is this something we can "fix" inside Bundler instead, so it's not necessary?

I've thought about this a decent amount. My possible ideas:
 * First Idea:
   * Have a plugin hook to say "my plugin extends the DSL",
   * During the plugin-only evaluation of the Gemfile, if a plugin isn't installed yet, modify `plugin` method in two ways:
     * If the plugin isn't installed yet, exit the entire evaluation immediately and install just that plugin, then retry
     * If the plugin is installed and is registered to extend the DSL, load it immediately at this point
   * If you're updating a plugin though, we won't know you're out of date during this phase, and you could be using a new feature from the updated plugin that hasn't been installed yet, and it could still fail.
 * Essentially capture and ignore any exception during Gemfile evaluation for the plugin phase. You'd think it would just be `NoMethodError` (and possibly `LoadError` and `ArgumentError`), but I imagine it would be possible for just about anything to happen inside a plugin's DSL addition. This sounds kind of icky at first, but any "real" errors _should_ crop up again and be reported when we get to the primary evaluation of the Gemfile.

If one of these were implemented, it might be possible to eliminate the `validate_plugins!` from this PR. It's hard to know for sure until it's implemented. I'd definitely want to work on it as a separate PR.

> since the current error at least mentions that the path does not exist while the new error mentions that the plugin "is not installed", which does not make a lot of sense

🤔 and here I am thinking "this is more clear, because regardless of if the gem exists, it may not be "installed" into the plugin index in current working directory as a Bundler plugin". But that comes from the perspective of someone familiar with how plugins work in Bundler. So yeah, I guess it could be more detailed and/or conditional: "The plugin is not installed" if the directory doesn't exist, or "The gem for plugin X is installed, but is not registered with Bundler; please run `bundle install`"?

Reply:
Alright, that sounds good, let's give a more detailed message in that case and this is good to go from my side. We can work on improving the same error for standard gems later.

Reply:
Very interested in this conversation.  I have this exact problem with [bundler-inject](https://github.com/ManageIQ/bundler-inject), because it also extends the DSL. I have to have the callers do a weird patching method, and a new hook for adding DSL methods would be amazing. Here's the details I wrote up if you're interested: https://github.com/ManageIQ/bundler-inject?tab=readme-ov-file#what-is-this-sorcery

Reply:
@ccutrer Are you planning on providing the better error message you suggested? If you don't have time, I'm happy to give it a try myself, all these improvements have been delayed too long already :)